sw/device/lib/dif/dif_uart.c

```diff
@@ -0,0 +1,411 @@
+  return mmio_region_get_bit32(uart->base_addr, UART_STATUS(0),
+                               UART_STATUS_TXEMPTY);
+}
+
+static bool uart_rx_empty(const dif_uart_t *uart) {
+  return mmio_region_get_bit32(uart->base_addr, UART_STATUS(0),
+                               UART_STATUS_RXEMPTY);
+}
+
+static uint8_t uart_rx_fifo_read(const dif_uart_t *uart) {
+  uint32_t reg = mmio_region_read32(uart->base_addr, UART_RDATA(0));
+
+  return reg & UART_RDATA_MASK;
+}
+
+static void uart_tx_fifo_write(const dif_uart_t *uart, uint8_t byte) {
+  uint32_t reg = (byte & UART_WDATA_MASK) << UART_WDATA_OFFSET;
+  mmio_region_write32(uart->base_addr, UART_WDATA(0), reg);
+}
+
+/**
+ * Get the corresponding interrupt register bit offset. INTR_STATE, INTR_ENABLE
+ * and INTR_TEST registers have same bit offsets, so this routine can be reused.
+ */
+static ptrdiff_t uart_irq_offset_get(dif_uart_interrupt_t irq_type) {
+  ptrdiff_t offset;
+  switch (irq_type) {
+    case kDifUartInterruptTxWatermark:
+      offset = UART_INTR_STATE_TX_WATERMARK;
+      break;
+    case kDifUartInterruptRxWatermark:
+      offset = UART_INTR_STATE_RX_WATERMARK;
+      break;
+    case kDifUartInterruptTxEmpty:
+      offset = UART_INTR_STATE_TX_EMPTY;
+      break;
+    case kDifUartInterruptRxOverflow:
+      offset = UART_INTR_STATE_RX_OVERFLOW;
+      break;
+    case kDifUartInterruptRxFrameErr:
+      offset = UART_INTR_STATE_RX_FRAME_ERR;
+      break;
+    case kDifUartInterruptRxBreakErr:
+      offset = UART_INTR_STATE_RX_BREAK_ERR;
+      break;
+    case kDifUartInterruptRxTimeout:
+      offset = UART_INTR_STATE_RX_TIMEOUT;
+      break;
+    case kDifUartInterruptRxParityErr:
+      offset = UART_INTR_STATE_RX_PARITY_ERR;
+      break;
+    default:
+      __builtin_unreachable();
+  }
+
+  return offset;
+}
+
+/**
+ * Performs fundamental UART configuration.
+ */
+static bool uart_configure(const dif_uart_t *uart,
+                           const dif_uart_config_t *config) {
+  if (config->baudrate == 0 || config->clk_freq_hz == 0) {
+    return false;
+  }
+
+  // Calculation formula: NCO = 2^20 * baud / fclk
+  uint64_t nco = ((uint64_t)config->baudrate << 20) / config->clk_freq_hz;
+  uint32_t nco_masked = nco & UART_CTRL_NCO_MASK;
+
+  // Requested baudrate is too high for the given clock frequency
+  if (nco != nco_masked) {
+    return false;
+  }
+
+  // Set baudrate, enable RX and TX, configure parity
+  uint32_t reg = (nco_masked << UART_CTRL_NCO_OFFSET);
+  reg |= (1u << UART_CTRL_TX);
+  reg |= (1u << UART_CTRL_RX);
+  if (config->parity_enable == kDifUartEnable) {
+    reg |= (1u << UART_CTRL_PARITY_EN);
+  }
+  if (config->parity == kDifUartParityOdd) {
+    reg |= (1u << UART_CTRL_PARITY_ODD);
+  }
+  mmio_region_write32(uart->base_addr, UART_CTRL(0), reg);
+
+  // Reset RX/TX FIFOs
+  reg = (1u << UART_FIFO_CTRL_RXRST);
+  reg |= (1u << UART_FIFO_CTRL_TXRST);
+  mmio_region_write32(uart->base_addr, UART_FIFO_CTRL(0), reg);
```

Comment:
@eunchan @tjaychen Is it possible to write a `FIFO_CTRL_RXRST/TXRST` before writing `UART_CTRL_TX` and `UART_CTRL_RX` to clear the FIFOs before enabling TX and RX?

This doesn't block landing this, I see you're using the init sequence right from the `hw/ip/uart` reference, but I'm curious if TX/RX could be disabled while there was still data in the FIFO which would show up immediately after enabling TX/RX but before clearing the FIFOs.